cleanrl/ppo_mujoco_envpool_xla_jax.py

```diff
@@ -0,0 +1,679 @@
+    def step_once(carry, step, env_step_fn):
+        agent_state, episode_stats, obs, done, truncated, key, handle = carry
+        action, logprob, value, key = get_action_and_value(agent_state, obs, key)
+        (
+            episode_stats,
+            handle,
+            (next_obs, reward, next_done, next_truncated, _),
+        ) = env_step_fn(episode_stats, handle, action.astype(jnp.float64))
+        next_obs = next_obs.astype(jnp.float32)
+        reward.astype(jnp.float32)
+        storage = Storage(
+            obs=obs,
+            actions=action,
+            logprobs=logprob,
+            dones=done,
+            truncated=truncated,
+            values=value,
+            rewards=reward,
+            returns=jnp.zeros_like(reward),
+            advantages=jnp.zeros_like(reward),
+        )
+        return (
+            (
+                agent_state,
+                episode_stats,
+                next_obs,
+                next_done,
+                next_truncated,
+                key,
+                handle,
+            ),
+            storage,
+        )
+
+    def rollout(
+        agent_state,
+        episode_stats,
+        next_obs,
+        next_done,
+        next_truncated,
+        key,
+        handle,
+        step_once_fn,
+        max_steps,
+    ):
+        (agent_state, episode_stats, next_obs, next_done, next_truncated, key, handle,), storage = jax.lax.scan(
+            step_once_fn,
+            (
+                agent_state,
+                episode_stats,
+                next_obs,
+                next_done,
+                next_truncated,
+                key,
+                handle,
+            ),
+            (),
+            max_steps,
+        )
+        return (
+            agent_state,
+            episode_stats,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+            handle,
+        )
+
+    rollout = partial(
+        rollout,
+        step_once_fn=partial(step_once, env_step_fn=step_env_wrappeed),
+        max_steps=args.num_steps,
+    )
+    for update in range(1, args.num_updates + 1):
+        update_time_start = time.time()
+        (
+            agent_state,
+            episode_stats,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+            handle,
+        ) = rollout(agent_state, episode_stats, next_obs, next_done, next_truncated, key, handle)
+        global_step += args.num_steps * args.num_envs
+        agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key = update_ppo(
+            agent_state,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+        )
+        print(f"pg_loss={pg_loss.mean()}, loss={loss.mean()}, v_loss={v_loss.mean()}, entropy_loss={entropy_loss.mean()}")
```

Comment:
Please move the logging down. From the reference `ppo_continuous_action.py`, `pg_loss` should come after `v_loss`